lib/ash/type/union.ex

```diff
@@ -377,4 +377,67 @@ defmodule Ash.Type.Union do
       :error
     end
   end
+
+  @impl true
+  def handle_change(nil, %Ash.Union{type: type_name, value: new_value}, constraints) do
+    with {:ok, type_configs} <- Keyword.fetch(constraints, :types),
+         {:ok, type_config} <- Keyword.fetch(type_configs, type_name),
+         {:ok, type} <- Keyword.fetch(type_config, :type),
+         {:ok, type_constraints} <- Keyword.fetch(type_config, :constraints),
+         {:ok, new_value} <- type.handle_change(nil, new_value, type_constraints) do
+      {:ok, %Ash.Union{type: type_name, value: new_value}}
+    end
+  end
+
+  def handle_change(%Ash.Union{type: type_name, value: old_value}, nil, constraints) do
+    with {:ok, type_configs} <- Keyword.fetch(constraints, :types),
+         {:ok, type_config} <- Keyword.fetch(type_configs, type_name),
+         {:ok, type} <- Keyword.fetch(type_config, :type),
+         {:ok, type_constraints} <- Keyword.fetch(type_config, :constraints),
+         {:ok, new_value} <- type.handle_change(old_value, nil, type_constraints) do
+      {:ok, %Ash.Union{type: type_name, value: new_value}}
+    end
+  end
+
+  def handle_change(
+        %Ash.Union{type: type_name, value: old_value},
+        %Ash.Union{type: type_name, value: new_value},
+        constraints
+      ) do
+    with {:ok, type_configs} <- Keyword.fetch(constraints, :types),
+         {:ok, type_config} <- Keyword.fetch(type_configs, type_name),
+         {:ok, type} <- Keyword.fetch(type_config, :type),
+         {:ok, type_constraints} <- Keyword.fetch(type_config, :constraints),
+         {:ok, new_value} <- type.handle_change(old_value, new_value, type_constraints) do
+      {:ok, %Ash.Union{type: type_name, value: new_value}}
+    end
+  end
+
+  def handle_change(
+        %Ash.Union{value: old_value},
+        %Ash.Union{type: type_name, value: new_value},
+        constraints
+      ) do
+    with {:ok, type_configs} <- Keyword.fetch(constraints, :types),
+         {:ok, type_config} <- Keyword.fetch(type_configs, type_name),
+         {:ok, type} <- Keyword.fetch(type_config, :type),
+         {:ok, type_constraints} <- Keyword.fetch(type_config, :constraints),
+         {:ok, new_value} <- type.handle_change(old_value, new_value, type_constraints) do
+      {:ok, %Ash.Union{type: type_name, value: new_value}}
+    end
+  end
+
+  @impl true
+  # FIXME I don't really understand what to do with other ways this function is
```

Comment:
This one is essentially exactly the same as the `handle_change`, except we don't know the type of the new value here. It could be an `%Ash.Union{}` *or* it could be any value that will (or won't) cast to this type. So what we need to do is something like this:

```elixir
def prepare_change(_old_value, nil, _) do
  {:ok, nil}
end

def prepare_change(%Ash.Union{type: type_name}, new_value, constraints) do
  case cast_input(new_value, constraints) do
     {:ok, %Ash.Union{type: ^type_name, value: value}} ->
       do_prepare_change(other_type, nil, value, constraints)
     {:ok, %Ash.Union{type: other_type, value: value}} ->
        do_prepare_change(other_type, nil, value, constraints)
     {:error, _} ->
        {:ok, new_value}
   end
end

defp do_prepare_change(type_name, old_value, new_value, constraints) do
  # call `prepare_change` on the right type, and return `{:ok, %Ash.Union{type_name: type_name, value: prepared_value}}` on success
end
```